docs/src/tutorial.md

```diff
@@ -0,0 +1,197 @@
+# Tutorial
+
+NLPModelsIpopt is a thin IPOPT wrapper for NLPModels. In this tutorial we'll show a few
+examples of problems created with NLPModels and solved with Ipopt.
+
+```@contents
+Pages = ["tutorial.md"]
+```
+
+## Simple problems
+
+The interface for calling Ipopt is very simple:
+
+```@docs
+ipopt
+```
+
+Let's create an NLPModel for the Rosenbrock function
+```math
+f(x) = (x_1 - 1)^2 + 100 (x_2 - x_1^2)^2
+```
+to test this interface:
+
+```@example ex1
+using NLPModels, NLPModelsIpopt
+
+nlp = ADNLPModel(x -> (x[1] - 1)^2 + 100 * (x[2] - x[1]^2)^2, [-1.2; 1.0])
+stats = ipopt(nlp)
+print(stats)
+```
+
+For comparison, we present the same problem and output using the JuMP route:
+
+```@example ex2
+using JuMP, Ipopt
+
+model = Model(with_optimizer(Ipopt.Optimizer))
+x0 = [-1.2; 1.0]
+@variable(model, x[i=1:2], start=x0[i])
+@NLobjective(model, Min, (x[1] - 1)^2 + 100 * (x[2] - x[1]^2)^2)
+optimize!(model)
+```
+
+Another example, using a constrained problem
+
+```@example ex1
+n = 10
+x0 = ones(n)
+x0[1:2:end] .= -1.2
+nlp = ADNLPModel(x -> sum((x[i] - 1)^2 + 100 * (x[i+1] - x[i]^2)^2 for i = 1:n-1), x0,
+                 c=x -> [3 * x[k+1]^3 + 2 * x[k+2] - 5 + sin(x[k+1] - x[k+2]) * sin(x[k+1] + x[k+2]) +
+                         4 * x[k+1] - x[k] * exp(x[k] - x[k+1]) - 3 for k = 1:n-2],
+                 lcon=zeros(n-2), ucon=zeros(n-2))
+stats = ipopt(nlp, print_level=0)
+print(stats)
+```
+
+## Output
+
+The output of `ipopt` is a `GenericExecutionStats` from `SolverTools`. It contains a few basic information of the solver.
+In addition to the built-in fields of `GenericExecutionStats`, we also store in
+`solver_specific` the following fields:
+
+- `multipliers_con`: Constraints multipliers;
+- `multipliers_L`: Variables lower-bound multipliers;
+- `multipliers_U`: Variables upper-bound multipliers;
+- `internal_msg`: Detailed Ipopt output message.
+
+```@example ex1
+stats.solver_specific[:internal_msg]
+```
+
+## Manual input
+
+This is an example of manually inputting the problem and its derivatives. For this,
+we're creating an NLPModel, and we need to define the following API functions:
+- `obj(nlp, x)`: objective
+- `grad!(nlp, x, g)`: gradient
+- `cons!(nlp, x, c)`: constraints, if any
+- `rows, cols = jac_structure(nlp)`: structure of the Jacobian, if constrained;
+- `jac_coord!(nlp, x, rows, cols, vals)`: Jacobian values (don't access `rows` and
+  `cols`, as Ipopt doesn't actually pass them;
```

Comment:
Should we add a layer of protection against this, and only ask the user to implement `jac_coord!(nlp, x, vals)`?

Reply:
Sorry, I didn't understand this, could you elaborate?

Reply:
If the user tries to access `rows` or `cols`, it causes a segfault, doesn't it? If so, we could only ask the user to implement `jac_coord!(nlp, x, vals)`, so that they never see `rows` or `cols`.

Reply:
Would we change `NLPModelsIpopt` to use `jac_coord!(nlp, x, vals)`? In this case, I image we can remove `jac_coord!(nlp, x, rows, cols, vals)`, though it would require changes to CUTEst (that requires `rows` and `cols`). We can keep `j_rows` and `j_cols` inside the struct of CUTEst in this case. This would be breaking, though, and would need to be in the next release of `NLPModels`.

Otherwise, we can define `jac_coord!(nlp, x, vals)`, and redefine `jac_coord!(nlp, x, rows, cols, vals)` to use `jac_coord!(nlp, x, vals)` by default. This would be somewhat ugly because it would keep two APIs for `jac_coord!`, but it doesn't seem breaking.

Reply:
I'm not sure what's best. We can save this for later. I was just trying to prevent accidental segfaults.